File: backend-python/utils/match_status.py
from datetime import datetime, timezone
from typing import Awaitable, Callable, Optional
from db import MatchStatus, Match

def get_match_status(
    start_time: Optional[datetime],
    end_time: Optional[datetime],
    now: Optional[datetime] = None
) -> Optional[MatchStatus]:
    """
    Determines the current match status based on time boundaries.
    """
    if now is None:
        now = datetime.now()
    if not start_time or not end_time:
        return None
    start_time = start_time.replace(tzinfo=None)
    if now < start_time:
        return MatchStatus.SCHEDULED
    end_time = end_time.replace(tzinfo=None)
    if now >= end_time:
        return MatchStatus.FINISHED
    return MatchStatus.LIVE

async def sync_match_status(
    match: Match,
    update_status: Callable[[MatchStatus], Awaitable[None]]
) -> MatchStatus:
    next_status = get_match_status(match.start_time, match.end_time)
    if not next_status:
        return match.status
    if match.status != next_status:
        await update_status(next_status)
        match.status = next_status
    return match.status

File: backend/src/utils/match_status.rs
use chrono::{DateTime, Utc};

use crate::entity::matches;
use crate::entity::sea_orm_active_enums::MatchStatus;

pub fn get_match_status(start: DateTime<Utc>, end: DateTime<Utc>) -> MatchStatus {
    let now = Utc::now();
    if now < start {
        MatchStatus::Scheduled
    } else if now > end {
        MatchStatus::Finished
    } else {
        MatchStatus::Live
    }
}

pub fn sync_match_status<F>(existing: &matches::Model, update_fn: F) 
where 
    F: FnOnce(MatchStatus) {
    let start_time = existing.start_time.and_utc();
    let end_time = existing.end_time.and_utc();
    let next_status = get_match_status(start_time, end_time);
    if next_status != existing.status {
        update_fn(next_status);
    }
}